### PR TITLE
GH-91079: Rename C_RECURSION_LIMIT to Py_C_RECURSION_LIMIT

### PR DESCRIPTION
**Pull request by vstinner**
Symbols of the C API should be prefixed by "Py_" to avoid conflict with existing names in 3rd party C extensions on #include <Python.h>.

* Issue: gh-91079

### CODE REVIEW

**Comment by vstinner** [comment]
@iritkatriel: Are you ok with this rename?

**Comment by vstinner** [comment]
cc @markshannon

File: Include/cpython/pystate.h [hunk 1]
```diff
@@ -197,12 +197,12 @@ struct _ts {
 /* WASI has limited call stack. Python's recursion limit depends on code
    layout, optimization, and WASI runtime. Wasmtime can handle about 700
    recursions, sometimes less. 500 is a more conservative limit. */
-#ifndef C_RECURSION_LIMIT
+#ifndef Py_C_RECURSION_LIMIT
```
Comment (iritkatriel) [inline Include/cpython/pystate.h]: Do users define this?

Comment (vstinner) [reply Include/cpython/pystate.h]: I don't know if it's relevant to support that. It's supposed to be a constant built into Python, no?

**Comment by vstinner** [comment]
@iritkatriel: I removed ``#ifndef Py_C_RECURSION_LIMIT``. Would you mind to review again the PR?

**Comment by vstinner** [comment]
> vstinner removed the needs backport to 3.12

It's too late to backport to Python 3.12.

**Comment by vstinner** [comment]
I don't think that it makes sense to override the ``Py_C_RECURSION_LIMIT`` constant. Python does hardcode this constant in its code. For example, ``init_threadstate()`` sets ``tstate->c_recursion_remaining = Py_C_RECURSION_LIMIT;``.

Building Python with a different ``Py_C_RECURSION_LIMIT`` value sounds like a borderline usecase. If someone cares about fine tuning ``Py_C_RECURSION_LIMIT``, I would suggest adding a ``./configure`` option. Currently, building Python with a different ``C_RECURSION_LIMIT`` value leads to inconsistencies for C extensions built with the default ``Py_C_RECURSION_LIMIT`` value (1500).

Including ``<Python.h>`` *and* overriding ``Py_C_RECURSION_LIMIT`` sounds like a bad idea: Python will anyway uses its hardcoded constants (ex: in ``init_threadstate()``), changing the constant doesn't change Python behavior. It only leads to inconsistency.

In short, IMO it's a good thing to remove ``#ifndef C_RECURSION_LIMIT`` :-)

**Review by iritkatriel — APPROVED** [review]

**Comment by vstinner** [comment]
Since Py_C_RECURSION_LIMIT value is different on WASI, I modified ``test.pythoninfo`` to log the value, to ease debugging test failures related to recursion. Recently, I added ``sys.getrecursionlimit`` to test.pythoninfo when I fixed test_tomllib recursion error (commit 8ff11425783806f8cb78e99f667546b1f7f3428e).
